pkg/cloudprovider/aws/fleet/capacity.go

```diff
@@ -0,0 +1,183 @@
+			TotalTargetCapacity:       aws.Int64(1), // TODO construct this more intelligently
+		},
+		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{{
+			LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
+				LaunchTemplateName: launchTemplate.LaunchTemplateName,
+				Version:            aws.String("$Default"),
+			},
+			Overrides: []*ec2.FleetLaunchTemplateOverridesRequest{{
+				AvailabilityZone: aws.String(zone),
+				InstanceType:     aws.String("m5.large"), // TODO construct this more intelligently
+				SubnetId:         subnet.SubnetId,
+			}},
+		}},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating fleet, %w", err)
+	}
+
+	// 4. Transform to Nodes.
+	var instanceIds []*string
+	for _, fleetInstance := range createFleetOutput.Instances {
+		instanceIds = append(instanceIds, fleetInstance.InstanceIds...)
+	}
+	nodes, err := c.nodeFactory.For(ctx, instanceIds)
+	if err != nil {
+		return nil, fmt.Errorf("determining nodes, %w", err)
+	}
+	zap.S().Infof("Successfully requested %d nodes", len(nodes))
+	return nodes, nil
+}
+
+// GetTopologyDomains returns a set of supported domains.
+// e.g. us-west-2 -> [ us-west-2a, us-west-2b ]
+func (c *Capacity) GetTopologyDomains(ctx context.Context, key cloudprovider.TopologyKey) ([]string, error) {
+	switch key {
+	case cloudprovider.TopologyKeyZone:
+		zones, err := c.getZones(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return zones, nil
+	case cloudprovider.TopologyKeySubnet:
+		subnets, err := c.getSubnetIds(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return subnets, nil
+	default:
+		return nil, fmt.Errorf("unrecognized topology key %s", key)
+	}
+}
+
+// seletZone chooses a zone for the given constraints.
+func (c *Capacity) selectZone(ctx context.Context, constraints *cloudprovider.CapacityConstraints) (string, error) {
+	// 1. Return zone if specified.
+	if zone, ok := constraints.Topology[cloudprovider.TopologyKeyZone]; ok {
+		return zone, nil
+	}
+	// 2. Randomly choose from available zones.
+	zones, err := c.getZones(ctx)
+	if err != nil {
+		return "", err
+	}
+	return zones[rand.Intn(len(zones))], nil
+}
+
+// selectSubnet chooses a subnet for the given constraints.
+func (c *Capacity) selectSubnet(ctx context.Context, zone string, constraints *cloudprovider.CapacityConstraints) (*ec2.Subnet, error) {
+	// 1. Get all subnets
+	zonalSubnets, err := c.subnetProvider.Get(ctx, c.spec.Cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. Return specific subnet if specified.
+	if subnetId, ok := constraints.Topology[cloudprovider.TopologyKeySubnet]; ok {
+		for _, subnets := range zonalSubnets {
+			for _, subnet := range subnets {
+				if subnetId == *subnet.SubnetId {
+					return subnet, nil
+				}
+			}
+		}
+		return nil, fmt.Errorf("no subnet exists named %s", subnetId)
+	}
+
+	// 3. Return random subnet in the zone.
+	subnets, ok := zonalSubnets[zone]
```

Comment:
Does having this only allow an even distribution of instances over all subnets? Is it possible (will a user even want this?) for a user to tweak this distribution?

Reply:
If a pod doesn't specify an AZ for any reason, we randomize.

Reply:
I see your point. I think this should be expressed via pod specs. I could imagine defaulting to a different strategy instead of random, though.